R/grid_code_paths.R

```diff
@@ -1,26 +1,161 @@
+    rng = rng,
+    parallel_over = parallel_over
+  )
+}
+
+parallel_over_finalize <- function(parallel_over, n_resamples) {
+  # Always use user supplied option, even if not as efficient
+  if (!is.null(parallel_over)) {
+    return(parallel_over)
+  }
+
+  # Generally more efficient to parallelize over just resamples,
+  # but if there is only 1 resample we instead parallelize over
+  # "everything" (resamples and the hyperparameter grid) to maximize
+  # core utilization
+  if (n_resamples == 1L) {
+    "everything"
+  } else {
+    "resamples"
+  }
+}
+
+# ------------------------------------------------------------------------------
+# Agua / h2o specific utilities
+
+tune_grid_loop_agua <- function(resamples,
+                                grid,
+                                workflow,
+                                metrics,
+                                control,
+                                rng) {
+  if (!rlang::is_installed("agua")) {
+    rlang::abort("`agua` must be installed to use an h2o parsnip engine.")
+  }
+
+  if (!is.null(grid) && !is_regular_grid(grid)) {
+    rlang::abort("The h2o engine only supports regular tuning grids.")
+  }
+
+  parallel_over <- control$parallel_over
+  parallel_over <- parallel_over_finalize_agua(parallel_over)
+
+  fn_tune_grid_loop_iter <- utils::getFromNamespace(
+    x = "tune_grid_loop_iter_h2o",
+    ns = "agua"
+  )
+
+  tune_grid_loop_impl(
+    fn_tune_grid_loop_iter = fn_tune_grid_loop_iter,
+    resamples = resamples,
+    grid = grid,
+    workflow = workflow,
+    metrics = metrics,
+    control = control,
+    rng = rng,
+    parallel_over = parallel_over
+  )
+}
+
+parallel_over_finalize_agua <- function(parallel_over) {
+  if (is.null(parallel_over)) {
+    parallel_over <- "resamples"
+  }
+
+  if (!identical(parallel_over, "resamples")) {
+    rlang::abort("`agua` only supports `parallel_over = \"resamples\".")
+  }
+
+  parallel_over
+}
+
+workflow_uses_agua <- function(workflow) {
+  model_spec <- hardhat::extract_spec_parsnip(workflow)
+  identical(model_spec$engine, "h2o")
+}
+
+# ------------------------------------------------------------------------------
+
+tune_grid_loop_impl <- function(fn_tune_grid_loop_iter,
+                                resamples,
+                                grid,
+                                workflow,
+                                metrics,
+                                control,
+                                rng,
+                                parallel_over) {
+  splits <- resamples$splits
+  packages <- c(control$pkgs, required_pkgs(workflow))
+  grid_info <- compute_grid_info(workflow, grid)
+
+  n_splits <- length(splits)
+  n_grid_info <- nrow(grid_info)
+
+  # Must assign `tune:::tune_grid_loop_iter_safely()` to a function in
+  # this function environment so foreach can find it. Mainly an issue with
+  # doParallel and PSOCK clusters.
+  fn_tune_grid_loop_iter_safely <- tune_grid_loop_iter_safely
```

Comment:
if returning a function, could use something like
`fn_tune_grid_loop_iter <- make_safely(fn_tune_grid_loop_iter)`, and then use `fn_tune_grid_loop_iter` in foreach